### DISCUSSION TITLE
BluePill CDC Serial only works w/ Arduino IDE

### DISCUSSION THREAD

**Discussion by nathantsoi** [stm32duino/Arduino_Core_STM32#2063] [Q&A] [answered]
I've been trying to get the USB CDC Serial working on a bluepill (stm32f103 128k), but I can only get it to work w/ the Arduino IDE. I've tried both PlatformIO and Cmake, but neither of these build systems will enable the CDC Serial Port.

I flashed the [DFU bootloader](https://github.com/rogerclarkmelbourne/STM32duino-bootloader/raw/master/binaries/generic_boot20_pc13.bin).

The working Arduino config looks like this:
<img width="509" alt="Screenshot 2023-07-10 at 9 16 06 AM" src="https://github.com/stm32duino/Arduino_Core_STM32/assets/399770/d54b002c-7ca6-4c36-a8d4-2656b0f3d966">

The resulting verbose build, including flags, is here: https://gist.github.com/nathantsoi/10359f187b604c0c29f26face55c4abb

I understand the build flags for PlatformIO should be:

```
build_flags = 
	-D PIO_FRAMEWORK_ARDUINO_ENABLE_CDC
	-D USBCON
```

However, this does not create the CDC Serial port.

I also tried with the cmake build system, with this configuration:

```
include(set_board)
set_board("BLUEPILL_F103CB"
  # SERIAL generic
  USB CDC
  # XUSB FS
  # VIRTIO disabled
  # BOOTLOADER dfuo / dfu2 / hid
)
```

Which also does not result in the activation of the CDC Serial port.

I checked the outputs of all the builds and made sure the relevant flags exist in all cases, such as:

```
-DUSBCON 
-DUSBD_VID=0 
-DUSBD_PID=0 
-DHAL_PCD_MODULE_ENABLED 
-DUSBD_USE_CDC 
-DHAL_UART_MODULE_ENABLED 
```

I'm happy to debug, but I'm not sure what I'm missing. Any direction is much appreciated.

Also, note that am using the `maple_upload.sh` script from the Arduino.app to ensure this is not an issue w/ uploading the binary to the board.

**Comment by fpistm** [comment]
Hi @nathantsoi 
First we do not use PIO nor support it so can't help on it.
For the CMake part, try:
```
include(set_board)
set_board("BLUEPILL_F103CB"
  # SERIAL generic
  USB CDCgen
  # XUSB FS
  # VIRTIO disabled
  # BOOTLOADER dfuo / dfu2 / hid
)
```

In fact `USB CDCgen` specify it is the default `Serial` instance (Serial == SerialUSB) while `USB CDC` only enable the SerialUSB and Serial is mapped to the default USART.

**Comment by nathantsoi** [comment]
Thanks for the response, @fpistm. the `USB CDC` -> `USB CDCgen` change seems to have no effect.

I also realized when testing this that perhaps there is a fault before the program code is even executed (which would explain why no serial port is starting). Even the simplest blink sketch won't run. I checked the linker params and see that the flash offset is not set correctly, nor is the `-DBL_LEGACY_LEAF` flag set ( per https://github.com/stm32duino/Arduino_Core_STM32/blob/main/boards.txt#L2876C60-L2876C76)

I tried manually setting these in the ninja linker config to no avail.

Any suggestions on how I might debug any faults before entering the application code?

Also, is [Stm32 cubeIDE](https://github.com/stm32duino/Arduino_Core_STM32/wiki/cubeIDE) the suggested debugger? I've only [used Eclipse](https://blog.nathantsoi.com/article/stm32-hardfault-debugging/).

**Reply by fpistm** [reply, marked as answer]
I've tested to create a project for AnalogReadSerial.ino and it works as expected. Here my CMakeFile.txt:

```cmake
# This file was autogenerated by ~/AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.6.0\cmake\scripts\cmake_easy_setup.py.
# Use it in your CMake configuration by `include()`'ing it.
# You can also copy it in your sketch's folder and edit it to fit your project.

cmake_minimum_required(VERSION 3.21)

# STEP 1: set up bases of environment
# -----------------------------------------------------------------------------

file(REAL_PATH "~/AppData\\Local\\Arduino15\\packages\\STMicroelectronics\\hardware\\stm32\\2.6.0" CORE_PATH EXPAND_TILDE)
file(TO_CMAKE_PATH "${CORE_PATH}" CORE_PATH)

file(REAL_PATH "C:\\IDE\\data\\Arduino\\libraries" USER_LIBS EXPAND_TILDE)
file(TO_CMAKE_PATH "${USER_LIBS}" USER_LIBS)

set(BOARDNAME "BLUEPILL_F103C8")

list(APPEND CMAKE_MODULE_PATH ${CORE_PATH}/cmake)
set(CMAKE_TOOLCHAIN_FILE toolchain)


# You may remove this block when using this file as the sketch's CMakeLists.txt
if (NOT ${CMAKE_PARENT_LIST_FILE} STREQUAL ${CMAKE_CURRENT_LIST_FILE})
    # When we are imported from the main CMakeLists.txt, we should stop here
    # not to interfere with the true build config.
    return()
endif()

project("AnalogReadSerial_project")

# STEP 2: configure the build
# -----------------------------------------------------------------------------

# Uncomment and pick the relevant value for each keyword!
# The first value listed is the default (when the feature is supported by the board)
# This means that leaving everything commented out yields the default config
include(set_board)
set_board("${BOARDNAME}"
  # SERIAL disabled
  USB CDCgen
  # XUSB FS / HS / HSFS
  # VIRTIO disable / generic / enabled
  # BOOTLOADER dfuo / dfu2 / hid
)

include(overall_settings)
overall_settings(
  # STANDARD_LIBC
  # PRINTF_FLOAT
  # SCANF_FLOAT
  # DEBUG_SYMBOLS
  # LTO
  # NO_RELATIVE_MACRO
  # UNDEF_NDEBUG
  # OPTIMIZATION "s"
  # BUILD_OPT ./build.opt
  # DISABLE_HAL_MODULES ADC I2C RTC SPI TIM DAC EXTI ETH SD QSPI
  # CORE_CALLBACK
)

# STEP 3: configure your sketch
# -----------------------------------------------------------------------------
include(external_library)
# I cannot tell the dependencies of the library ahead-of-time
# Please write them in using the DEPENDS ... clause
# The same directives apply as for `build_sketch()` just below.
# external_library(PATH "${USER_LIBS}/AccelStepper")

include(build_sketch)
build_sketch(TARGET "AnalogReadSerial"
  SOURCES
  AnalogReadSerial.ino

  # Uncomment the lines below to bind libraries to your sketch
  # Legitimate names after the DEPENDS keywords are:
  # - libraries declared with external_library
  # - libraries from the libraries/ folder of Arduino_Core_STM32

  # DEPENDS
  # SD
  # Wire
  # SPI
)

# STEP 4: optional features
# -----------------------------------------------------------------------------

include(insights)
insights(TARGET "AnalogReadSerial"
  # DIRECT_INCLUDES
  # TRANSITIVE_INCLUDES
  # SYMBOLS
  # ARCHIVES
  # LOGIC_STRUCTURE
)
```

Result of the build:
```
Sketch uses 24472 bytes (37%) of program storage space. Maximum is 65536 bytes.
Global variables use 4380 bytes (21%) of dynamic memory, leaving 16100 bytes for local variables. Maximum is 20480 bytes.
```

![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/20641798/3fe9dd08-cf32-4168-84d1-14bfb541c192)

**Comment by nathantsoi** [comment]
Thanks again @fpistm. I tried again, this time on Windows, thinking it could be a problem w/the build chain on mac, but had no luck flashing the board at all, with either the DFU maple method or my st-link.

How are you flashing the board, st-link? Are there build options I need to change depending on the method of flashing?

**Reply by fpistm** [reply]
only some upload method have impact on the ldscript to add an offset if you havec bootloader. Anyway with STLink no impact.
Try to connect to your board using the STM32CubeProgrammer.
